pyro/distributions/lkj.py

```diff
@@ -62,65 +60,55 @@ class _PartialCorrToCorrCholeskyTransform(Transform):
+            raise ValueError("PartialCorrToCorrCholesky transformation requires d choose 2 inputs.")
         D = int(D)
 
-        x = torch.zeros((D,D), device=z.device)
+        y = x.new_zeros(x.shape[:-1] + (D, D))
+        y[..., 0, 0] = 1
+        y[..., 1:, 0] = x[..., :(D - 1)]
 
-        x[0,0] = 1
-        x[1:,0] = current_x = z[:(D-1)]
-        i = D - 1
-        last_squared_x = None
+        pos_x = D - 1
+        past_y_squared_sum = x.new_zeros(x.shape[:-1] + (D,))
+        # FIX ME: find a vectorized way to compute y instead of loop
         for j in range(1, D):
-            distance_to_copy = D - 1 - j
-            new_z = z[i:(i + distance_to_copy)]
-            if last_squared_x is None:
-                last_squared_x = current_x**2
-            else:
-                last_squared_x = last_squared_x[1:] + current_x**2
-            x[j, j] = (1 - last_squared_x[0]).sqrt()
-            current_x = new_z * (1 - last_squared_x[1:]).sqrt()
-            x[(j+1):, j] = current_x
-            i += distance_to_copy
-
-        return x
-
-    def _inverse(self, x):
-        if (x.shape[0] != x.shape[1]):
-            raise ValueError("A matrix that isn't square can't be a Cholesky factor of a correlation matrix")
-        D = x.shape[0]
-
-        z_stack = [
-            x[1:, 0]
-        ]
-        current_x = z_stack[0]
-        last_squared_x = None
-        for j in range(1, D):
-            if last_squared_x is None:
-                last_squared_x = current_x**2
-            else:
-                last_squared_x += current_x[1:]**2
-        current_x = x[j:, j]
-        z_stack.append(current_x / (1 - last_squared_x).sqrt())
-        z = torch.cat(z_stack)
-        return z
+            # XXX we need `.clone()` operator here to avoid the in-place operator error
+            # in the backward pass
+            past_y_squared_sum = past_y_squared_sum[..., 1:] + y[..., j:, (j - 1)].clone() ** 2
+            y[..., j, j] = (1 - past_y_squared_sum[..., 0]).sqrt()
+            new_pos_x = pos_x + D - 1 - j
+            y[..., (j + 1):, j] = x[..., pos_x:new_pos_x] * (1 - past_y_squared_sum[..., 1:]).sqrt()
+            pos_x = new_pos_x
+        return y
 
-    def log_abs_det_jacobian(self, x, z):
-        return (1 - x.tril(-1).pow(2).sum(1)).log().sum() * .5
+    def _inverse(self, y):
+        x = y.new_ones(y.shape)
+        triu_index = x.triu(diagonal=1) > 0.5
+        x[..., :, 0] = y[..., :, 0]
+        x[..., :, 1:] = y[..., :, 1:] / (1 - y.pow(2).cumsum(-1)[..., :, :-1]).sqrt()
+        # we transpose and take upper triangular indices to arrange the result vector
+        # by (x21, x31, x41,..., x32, x42,...) instead of (x21, x31, x32, x41, x42,...)
+        return x.transpose(-1, -2)[triu_index].reshape(x.shape[:-2] + (-1,))
+
+    def log_abs_det_jacobian(self, x, y):
+        triu_index = y.new_ones(y.shape).triu(diagonal=1) > 0.5
```

Comment:
Is `.triu` going to operate correctly when there's one or more dimensions to the batch? I.e., if the input has shape `[4, 2, 3, 3]`, is `.triu` going to do what's expected? 

This approach with the indices also looks to be slower than the approach of just taking the triangle, flattening it, and summing it. 

Also, the transformation should be making the *lower* cholesky factor, not the upper one.  The reason is that the `multinormal` and related distributions all take a `scale_tril` parameter, which is the *lower* cholesky factor of the covariance matrix.

Reply:
Yes, triu support batch.

Reply:
I saw it, but I wasn't persuaded.  Also, do you see the point about needing to return a *lower* cholesky factor?

Reply:
I don't see what you mean, could you elaborate it, pls?

Reply:
I mean the transformation needs to be returning a cholesky factor in the *lower* triangle of the matrix, not the *upper* triangle. Because the result is going to be fed into distribution classes that take as a parameter the *lower* cholesky factor, not the *upper* one.

Reply:
BTW - I'm working on this now, and I'm trying to use/test your code for the inverse.  I'm having trouble with the call to `triu` not supporting batches:

```
>>> torch.randn(2, 3, 3).triu(diagonal=1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: invalid argument 1: expected a matrix at /Volumes/ShareData/LIBRARIES/pytorch/aten/src/TH/generic/THTensorMoreMath.cpp:1286
```

Am I missing something?  And would you prefer if I do a push where I leave inverse unimplemented so you can do it?

Reply:
I guess the reason you were worried about the performance/optimized code instead of focusing on correctness first is because you have your own version work. So just use your version. It’ll reduce your time. I’ll not make PRs to this branch anymore because I can’t follow your idea and I think it is ready (after you fix your errors and add your tests). I’ll optimize the code in the following PRs in pyro repository, where I can give benchmark examples to compare, rather than to discuss about effectiveness or ineffectiveness “without code” here.